Добавил проверку на наличие скоупа print.file.send

BombinBM opened the pull request:
В коде добавил опциональность для ввода номера билета, а так же проверку на наличие скоупа внутри имеющихся у юзера. в выводе оставил только имя юзера.

BombinBM committed:
Добавил проверку на наличие скоупа print.file.send

В коде добавил опциональность для ввода номера билета, а так же проверку на наличие скоупа внутри имеющихся у юзера. в выводе оставил только имя юзера.

Temmmmmo commented:
линтинг упал, пропиши 
`black .`
`isort .`

Temmmmmo commented:
тебе нужен был файл `file.py`
там нужно было разобраться с одной ручкой, отвечающей за отправку файлов
метод check_union_member тебе вероятно не нужен)

BombinBM committed:
Вернул check_union_member и изменил file_send

В POST запросе check_union_member вернул код в состояние до первого комита.
Для функции send написал одну дополнительную функцию, чтоб проверять наличие скоупа через Depends, добавил опциональный ввод поля number и проверку его наличия уже в самой функции send

BombinBM committed:
Сделал black . isort .

parfenovma commented on print_service/routes/file.py:
```diff
@@ -107,21 +113,36 @@ class ReceiveOutput(BaseModel):
 )
-async def send(inp: SendInput, settings: Settings = Depends(get_settings)):
+async def send(
+    inp: SendInput,
```
нет тайп хинта, непонятное название

parfenovma commented on print_service/routes/file.py:
```diff
@@ -107,21 +113,36 @@ class ReceiveOutput(BaseModel):
-async def send(inp: SendInput, settings: Settings = Depends(get_settings)):
+async def send(
+    inp: SendInput,
+    has_send_scope: bool = Depends(has_send_scope),
```
обычно булевые переменнные начинаются с is_ или заканчиваются _flg

parfenovma commented on print_service/routes/file.py:
```diff
@@ -107,21 +113,36 @@ class ReceiveOutput(BaseModel):
+    if not has_send_scope and inp.number is None:
+        raise HTTPException(
+            status_code=400,
+            detail="Поле number обязательно для пользователей без скоупа print.file.send",
```
не юзер френдли ошибка. Она же на фронт пойдет, зачем нам писать служебную информацию? 
И наверное это 403 всё-таки

parfenovma commented on print_service/routes/file.py:
```diff
@@ -107,21 +113,36 @@ class ReceiveOutput(BaseModel):
-        ),
-        func.upper(UnionMember.surname) == inp.surname.upper(),
-    ).one_or_none()
+    if inp.number is not None:
```
`if inp.number:`

parfenovma reviewed (dismissed):
лучше ветки на англ называть
![Uploading image.png…]()

BombinBM committed:
Заменил вывод ошибки

Temmmmmo commented on print_service/routes/file.py:
```diff
@@ -107,21 +113,33 @@ class ReceiveOutput(BaseModel):
+            ),
+            func.upper(UnionMember.surname) == inp.surname.upper(),
+        ).one_or_none()
+    else:
```
давай этот else уберем и реализуем иначе

Temmmmmo commented on print_service/routes/file.py:
```diff
@@ -107,21 +113,33 @@ class ReceiveOutput(BaseModel):
+        user = user.filter(
+            func.upper(UnionMember.surname) == inp.surname.upper(),
+        )
     if not user:
```
вот тут добавим также проверку на наличие скоупа

Temmmmmo replied:
то есть если у чела не прописаны ни номер билета, ни фио, ни студак, но при этом есть скоуп, то все хорошо, ошибку не рейзим

BombinBM replied:
if user is None

Temmmmmo commented on print_service/routes/file.py:
```diff
@@ -96,6 +98,10 @@ class ReceiveOutput(BaseModel):
 
 
 # endregion
+def has_send_scope(
```
давай не будем это выносить в отдельную функцию

Temmmmmo commented on print_service/routes/file.py:
```diff
@@ -107,21 +113,33 @@ class ReceiveOutput(BaseModel):
+async def send(
+    inp: SendInput,
+    has_send_scope: bool = Depends(has_send_scope),
+    settings: Settings = Depends(get_settings),
```
депендс пропишем прям тут, это норм

Temmmmmo replied:
депендс от юнионаутха я имею в виду

Temmmmmo replied:
назовем его user, тогда в дальнейшем оттуда мы сможем еще получать user_id

BombinBM replied:
auth_user

Temmmmmo commented on print_service/routes/file.py:
```diff
@@ -63,7 +64,8 @@ class SendInput(BaseModel):
         description='Фамилия',
```
смотри, surname тоже становится необязательным для пользователей со скоупом, давай учтем это

Temmmmmo commented on print_service/routes/user.py:
```diff
@@ -46,10 +46,11 @@ class UpdateUserList(BaseModel):
 async def check_union_member(
     surname: constr(strip_whitespace=True, to_upper=True, min_length=1),
-    number: constr(strip_whitespace=True, to_upper=True, min_length=1),
+    number: Optional[str] = constr(strip_whitespace=True, to_upper=True, min_length=1),
```
не уверен, что тут надо делать поле опциональным
т.к. для пользователей со скоупом в принципе не планируется прокидывание проверки на присутствие в профсоюзе

BombinBM committed:
Update file.py

Сделал все для ревью

BombinBM committed:
Update file.py

BombinBM commented on print_service/routes/file.py:
```diff
@@ -107,21 +109,34 @@ class ReceiveOutput(BaseModel):
-        func.upper(UnionMember.surname) == inp.surname.upper(),
-    ).one_or_none()
+
+    if inp.number is not None:
```
and inp.surname is not None

BombinBM commented on print_service/routes/file.py:
```diff
@@ -107,21 +109,34 @@ class ReceiveOutput(BaseModel):
+            func.upper(UnionMember.surname) == inp.surname.upper(),
+        ).one_or_none()
+    else:
+        if not "print.file.send" in [scope["name"] for scope in user.get('session_scopes')]:
```
elif

BombinBM committed:
Сделал все откомменченное

Осталось только разобраться с базой данных и тестами

BombinBM committed:
Локальные тесты + проверка скоупов

Сделал локальную проверку на наличие скоупа print.file.send
Пришлось вынести user = user.one_or_none() так как при разных условиях возвращаются разные типы данных (запрос в базу или UnionMember), а далее для создания пина нужен именно UnionMember.

Изменение в файле __main__.py предложил сделать Стас, чтоб автоматически при обновлениях кода обновлялось и приложение в вебе

Temmmmmo reviewed (commented):
Да вроде нормально
Ты тестил?

Temmmmmo reviewed (commented):
Да вроде нормально
Ты тестил?

Temmmmmo commented on print_service/routes/file.py:
```diff
@@ -107,22 +110,35 @@ class ReceiveOutput(BaseModel):
+
+    else:
+        if not "print.file.send" in [scope["name"] for scope in user_auth.get('session_scopes')]:
+            raise NotInUnion()
```
Единственное, возможно стоит ошибку тут другую кидать, иначе с ходу непонятно: это у тебя скоупа нет или тебя в профсоюзе нет

BombinBM replied:
Тестил локально, да.

Насчет ошибки не уверен. У нас же скоуп должен быть всегда, если человек состоит в профсоюзе

Temmmmmo commented on print_service/__main__.py:
```diff
@@ -1,7 +1,5 @@
 
 if __name__ == '__main__':
-    uvicorn.run(app)
+    uvicorn.run("print_service.routes.base:app", reload=True)
```
убрать

BombinBM replied:
Сделал

Temmmmmo reviewed (commented):
как будто норм, пофикси последний комм и мерджим

BombinBM committed:
Оставил только app в мейне

Сделал крайний коммент

Temmmmmo commented on print_service/__main__.py:
```diff
@@ -1,7 +1,5 @@
-
 
 if __name__ == '__main__':
     uvicorn.run(app)
```
Будет ошибка
app not found

BombinBM replied:
Да, добавил импорт. Теперь запускается и ручки работают

BombinBM committed:
Update __main__.py

Temmmmmo committed:
Merge branch 'main' into 80-добавление-аутха-в-принтер-как-второго-метода-авторизации
